### PR TITLE
Add Higgs audio model 

### LINKED ISSUE #32068: [Bounty $1500] Higgs Audio v2 bring up using TTNN APIs
### 📝 Background

This bounty is for bringing up the Higgs Audio v2 model using TTNN APIs on Tenstorrent hardware (Wormhole or Blackhole).
Higgs Audio v2 is a cutting-edge text-audio foundation model from Boson AI that redefines expressiveness in audio generation. Released in 2025, it offers state-of-the-art capabilities:
Exceptional expressiveness: Industry-leading performance on emotion and prosody benchmarks
Multi-speaker dialog: Natural multi-speaker conversations with distinct voices
Voice cloning: Zero-shot voice cloning from reference audio
Diverse audio generation: Speech, sound effects, music, and environmental sounds
Massive training scale: Trained on 10 million hours of audio data (AudioVerse dataset)
Custom tokenizer: Unified semantic and acoustic tokenization
DualFFN architecture: Enhanced LLM for acoustic modeling with minimal overhead
Apache 2.0 license: Fully open source and commercially usable
The model achieves 75.71% win-rate on EmergentTTS-Eval emotions and best-in-class multi-speaker generation quality.
The goal is to enable this model to run on TT hardware for high-throughput, low-latency audio generation across diverse use cases including virtual assistants, audiobooks, content creation, and interactive media.

### 🎯 What Success Looks Like

A successful submission will fulfill all requirements in the following stages. Payout is made after all three stages are completed.

### Stage 1 — Bring-Up

- Implement Higgs Audio v2 using TTNN APIs (Python)
- Implements the full generation pipeline:
         LLM backbone with DualFFN architecture
        Audio tokenizer (semantic + acoustic features)
        Audio decoder (token-to-waveform conversion)
- Model runs on either N150 or N300 tenstorrent hardware with no errors
- Supports multiple generation modes:
      Text-to-speech: Generate speech from text only
      Voice cloning: Generate speech with reference audio
      Multi-speaker dialog: Generate conversations with distinct speakers
- Produces valid audio output on sample texts (English and multilingual)
- Output is verifiable (audio quality assessment, compare with PyTorch reference)
- Achieves baseline throughput target:
      At least 30 tokens/second for autoregressive generation
      Real-time factor (RTF) < 0.5 for typical sentences
- Accuracy evaluation: Token-level accuracy > 95% against PyTorch reference
- Audio quality: Passes intelligibility and expressiveness tests
- Clear instructions for setup and running the model

### Stage 2 — Basic Optimizations

- Use optimal sharded/interleaved memory configs for LLM layers
- Implement efficient sharding strategy for:
    - Token embeddings (text + audio tokens)
    - DualFFN transformer layers
- Multi-head attention mechanisms
- Audio tokenizer encoder/decoder
- Fuse simple ops where possible (e.g., layer normalization, attention patterns, activation functions)
- Store intermediate activations in L1 where beneficial
- Use recommended TTNN/tt-metal LLM flows
- Leverage TT library of fused ops for attention and MLP blocks 
- Optimize DualFFN architecture (dual feed-forward networks)
- Efficient KV-cache management for autoregressive generation
- Optimize audio tokenizer integration

### Stage 3 — Deeper Optimization

- Maximize core counts used per inference
- Implement deeper TT-specific optimizations:
      - Efficient KV-cache management for long sequences
      - Optimized DualFFN computation (parallel FFN paths)
      - Flash Attention or equivalent for attention layers
- Minimize token generation latency
- Batch processing for multiple utterances/speakers
- Efficient sampling strategies (temperature, top-p, top-k)
- Pipeline audio encoding/decoding with LLM generation
- Minimize memory and TM (tensor manipulation) overheads
- Explore speculative decoding or other acceleration techniques
- Document any advanced tuning, known limitations, or trade-offs
- Target stretched goals:
        - 60+ tokens/second generation speed
        - RTF < 0.2 for real-time applications
- Support for longer contexts (multi-turn dialog)
- Efficient multi-speaker handling 

### 🧭 Guidance & Starting Points

* Use the [TTNN model bring-up tech report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md) as your primary reference
* Reference [LLM implementations in tt-metal](https://github.com/tenstorrent/tt-metal/tree/main/models/demos/wormhole) for LLM-based model patterns
* Use the [official Higgs Audio v2 repository](https://github.com/boson-ai/higgs-audio) for model architecture details
* Refer to the [Higgs Audio v2 blog](https://www.boson.ai/blog/higgs-audio-v2) for technical details
* Refer to [tokenizer blog](https://www.boson.ai/blog/audio-tokenizer) for audio tokenizer details
* Refer to [DualFFN architecture blog](https://www.boson.ai/blog/dualffn-architecture) for architecture innovations
* Refer to [TT Fused ops PR #29236](https://github.com/tenstorrent/tt-metal/pull/29236) for optimization opportunities
* The model architecture consists of:
  * **Audio tokenizer**: Unified semantic and acoustic feature extraction
  * **LLM backbone**: Transformer with DualFFN architecture
  * **DualFFN layers**: Dual feed-forward networks for text and audio tokens
  * **Audio decoder**: Token-to-waveform conversion (24 kHz output)
* Key challenges:
  * DualFFN architecture optimization (dual FFN paths)
  * Large vocabulary (text + audio tokens)
  * Autoregressive generation latency
  * Audio tokenizer integration
  * Multi-speaker context management
* Ask for help or file issues if ops are missing in TTNN

### 🔎 Possible Approaches

- Start from the [official Higgs Audio v2 repository](https://github.com/boson-ai/higgs-audio)  and port components sequentially:            
- End-to-end pipeline integration
- Validate each submodule's output against PyTorch reference before integration
- For example the LLM backbone:
- Start with standard transformer layers
- Add DualFFN architecture (key innovation)
- Optimize attention mechanisms
- Implement efficient KV-cache 
- Experiment with different sharding strategies
- Use TTNN profiling tools to identify bottlenecks 
- Test diverse use cases
- Open a draft PR early to get feedback on your approach

## 📊 Result Submission Guidelines

Beyond the model implementation itself, contributors must submit the following material as proof of work.
However, feel free to open a PR at any time if you want us checking that you are on the right track.
Just understand that payout is only made after **all 3 stages** are completed.

**Deliverables:**
- Functional model implementation
- Validation logs (output correctness)
- Performance report + header for final review

**Links:**
- [Performance Sheet](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md#41-performance-sheet)
- [Perf Header Docs](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/profiling_ttnn_operations.html#perf-report-headers)

### Model Resources
* [Higgs Audio v2 Official Repository](https://github.com/boson-ai/higgs-audio)
* [Higgs Audio v2 Release Blog](https://www.boson.ai/blog/higgs-audio-v2)
* [Audio Tokenizer Technical Blog](https://www.boson.ai/blog/audio-tokenizer)
* [DualFFN Architecture Blog](https://www.boson.ai/blog/dualffn-architecture)
* [Boson AI Website](https://www.boson.ai/)

### Evaluation Resources
* [Seed-TTS Eval Benchmark](https://github.com/ByteDance/seed-tts-eval)
* [Emotional Speech Dataset (ESD)](https://github.com/HLTSingapore/Emotional-Speech-Data)
* [EmergentTTS-Eval Paper](https://arxiv.org/abs/2410.01576)

### TT-Metal Resources
* [TTNN Model Bring-up Tech Report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md)
* [LLM Implementations in tt-metal](https://github.com/tenstorrent/tt-metal/tree/main/models/demos/wormhole)
* [TT Fused Ops PR #29236](https://github.com/tenstorrent/tt-metal/pull/29236)
* [Performance Report Header](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/profiling_ttnn_operations.html#perf-report-headers)

### PR DESCRIPTION
**Pull request by Jatin-exe**
### Ticket
#32068

### Problem description
This PR brings up a minimal public Higgs Audio v2 TTNN demo on Tenstorrent hardware.

- single-device Wormhole path
- `batch_size=1`
- greedy generation only
- three supported public modes: text-to-speech, voice cloning, and multi-speaker dialog
- two end-to-end hardware tests for correctness and performance

### Core Architecture

- TTNN-backed Higgs LLM backbone with DualFFN execution lives under `models/demos/audio/higgs_audio_v2/tt`.
- Public demo entrypoint is `models/demos/audio/higgs_audio_v2/demo/demo.py`.
- Supported public modes:
  - text-to-speech from transcript-only prompts
  - voice cloning from transcript + reference audio
  - ChatML multi-speaker dialog with pinned audio references
- Pinned reference-audio assets are managed through `fetch_reference_audio_assets.py` and `reference_audio_manifest.json`.
- The public runtime is fixed to single-device, `batch_size=1`, greedy decode.
- Prompt-audio encoding and final waveform decoding use the upstream Higgs audio tokenizer/decoder on host; the TTNN path covers the Higgs autoregressive backbone and device-side decode loop.

### Validation & Accuracy

- End-to-end Higgs test surface: `2/2` tests passed on Wormhole.
- `test_accuracy.py` runs TTNN against the PyTorch Higgs reference on three real cases:
  - multilingual TTS
  - voice cloning with reference audio
  - two-speaker dialog
- The accuracy gate enforced by the test is:
  - token-level accuracy against the PyTorch reference `>= 0.95`

Validation command:
```bash
python_env/bin/python -m pytest models/demos/audio/higgs_audio_v2/tests/test_accuracy.py -q
```

### Performance

- `test_performance.py` runs the traced TTNN performance path on three long-horizon cases matching the public demo modes.
- Current promoted public numbers from `models/demos/audio/higgs_audio_v2/PERF.md`:

Per-case numbers from the same promoted run:

| Case | Tok/s | RTF | Decode RTF |
|---|---:|---:|---:|
| `tts_es_project_update_long` | 123.14 | 0.20666 | 0.20302 |
| `clone_en_review_update_long` | 123.14 | 0.21297 | 0.20302 |
| `dialog_en_review_long` | 123.14 | 0.21300 | 0.20302 |
| aggregate | 123.14 | 0.21087 | 0.20302 |

Performance command:
```bash
python_env/bin/python -m pytest models/demos/audio/higgs_audio_v2/tests/test_performance.py -q
```

### Instructions

- export `TT_METAL_HOME`
- export `HIGGS_AUDIO_REPO`
- export `PYTHONPATH=$HIGGS_AUDIO_REPO:$TT_METAL_HOME${PYTHONPATH:+:$PYTHONPATH}`
- fetch pinned reference audio assets
- run one of the three public demo commands:
  - TTS
  - voice clone
  - dialog

### Optimisations

- TTNN Higgs backbone runs through the model code under `models/demos/audio/higgs_audio_v2/tt/model.py`.
- The performance path uses traced TTNN decode replay with persistent device decode-state buffers.

### Sample Demo Runs

| Mode | Input | Output artifact | 
| --- | --- | --- |
| TTS | `Please welcome everyone to the review and keep the tone calm and clear.` | [demo_tts.zip](https://github.com/user-attachments/files/26305590/demo_tts.zip)  |

### Known limitations

- Public runtime is greedy-only.
- Public runtime is single-device and `batch_size=1` only.
- Audio tokenizer encode/decode remains upstream host-side.
- This PR does not expose sampling controls such as `temperature`, `top-p`, or `top-k`.

### Checklist

- [ ] [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:{{branch_name}})
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:{{branch_name}})
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:{{branch_name}})
- [ ] New/Existing tests provide coverage for changes
- [ ] (Optional) Ran [clang-tidy code analysis](https://github.com/tenstorrent/tt-metal/actions/workflows/code-analysis.yaml) on the PR branch to catch linting errors early

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:{{branch_name}})
  - [ ] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:{{branch_name}})
  - [ ] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:{{branch_name}})
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1212819708891052

### CODE REVIEW

**Comment by jberkowitzTT** [comment]
Hi @Jatin-exe - were you able to satisfy the `RTF <= 0.2` requirement? If not, do you have an updated timeline for when that might be done?

**Comment by Jatin-exe** [comment]
Hi @jberkowitzTT 
I'm on it, Will update soon on the status
Thank you

**Comment by Jatin-exe** [comment]
Hi @jberkowitzTT - My TT instance does not seem to be working, can i mail you for access to Tenstorrent ? 
Thank you

**Comment by jberkowitzTT** [comment]
sure thing

email: jberkowitz@tenstorrent.com

**Comment by jberkowitzTT** [comment]
Hi @Jatin-exe - please provide updates & complete by May 27th. Otherwise, we'll reopen & reassign.

Thanks you, and let me know if I can help at all!

**Comment by jberkowitzTT** [comment]
Let me know if you're hitting issues or if there's anything else I can do to help!

**Comment by Jatin-exe** [comment]
Thank you @jberkowitzTT , I have achieved the target of <0.2 RTF

With this all the targets are achieved

I will be updating the PR asap

**Comment by Jatin-exe** [comment]
I'll be pushing the final changes as soon as I can (will be pushed by tmrw) 

Thank you

**Comment by Jatin-exe** [comment]
Hi @jberkowitzTT

The core bounty requirements were already completed and pushed earlier. Since then, I’ve been focused on the stretch optimization target, specifically achieving `RTF < 0.2`

That stretch goal has now also been completed and pushed to the repo. The PR is ready for review.

Thanks!

**Comment by jberkowitzTT** [comment]
Hi @Jatin-exe - perfect, thanks! I'll work with the engineering team to get the final review soon!

**Comment by Jatin-exe** [comment]
Looking into the failed test cases. 

My koyeb instance is erroring out. 

image

Trying to triage/troubleshoot this.

**Comment by Jatin-exe** [comment]
Hello @jberkowitzTT 
Can you run all of the tests again ? 
Thank you

**Comment by Jatin-exe** [comment]
The latest rebase seems to have broken a few things, I'm sending a fix

**Comment by jberkowitzTT** [comment]
Hi @Jatin-exe - unfortunately the final determination was that this does not meet the bounty requirements. I'll reopen this issue, but we deeply appreciate your efforts & look forward to your future contributions
